[#47043] Add storage links to the project sidebar menu

ba1ash opened the pull request:
https://community.openproject.org/work_packages/47043

ba1ash committed:
[#47043] Add storage links to the project sidebar menu

https://community.openproject.org/work_packages/47043

ba1ash committed:
Remove some unsused code.

ba1ash committed:
Extract storage menu modification to the module.

- Add `dynamic_menu` method to Redmine::Plugin.
- Fix tests.

ba1ash committed:
Rename new method for configuring menu in plugins.

ba1ash committed:
Add feature specs.

ba1ash committed:
Fix failing feature spec.

Click link directly, not a span inside it.

ba1ash commented on app/controllers/projects_controller.rb:
```diff
@@ -65,10 +65,6 @@ def index
     end
   end
 
-  current_menu_item :index do
```
It has been removed, because its a duplicate call. See the line 43.

ba1ash commented on app/controllers/projects_controller.rb:
```diff
@@ -109,12 +105,6 @@ def find_optional_project
     render_404
   end
 
-  def redirect_work_packages_or_overview
```
It has been removed, because it is no used.

ba1ash commented on lib/redmine/menu_manager.rb:
```diff
@@ -30,26 +30,16 @@ module Redmine::MenuManager
     @menu_builder_queues ||= {}
     current_queue = @menu_builder_queues[menu_name.to_sym] ||= []
-
-    if menu_builder
```
This condition has been simplified, because `.map` and  `.loose` method are called with block as a parameter. So, no need for `MapDeferrer` class as well.

ba1ash commented on lib/redmine/menu_manager/mapper.rb:
```diff
@@ -33,8 +33,6 @@ def initialize(menu, items)
     @menu_items = items[menu]
   end
 
-  @@last_items_count = Hash.new { |h, k| h[k] = 0 }
```
It is not used as well.

Kharonus commented on lib/redmine/menu_manager/menu_helper.rb:
```diff
@@ -322,7 +322,7 @@ def allowed_node?(node, user, project)
 
   def allowed_project_node?(node, project, user)
-    if node_action_allowed?(node, project, user)
+    if node_action_allowed?(node, project, user) || node.external_link?
```
why not putting the check `node.external_link?` into `node_action_allowed?` ? Seems like the better place for it.

Kharonus commented on modules/storages/lib/open_project/storages/engine.rb:
```diff
@@ -85,6 +85,24 @@ class Engine < ::Rails::Engine
            caption: :project_module_storages,
            parent: :settings
+
+      configure_menu :project_menu do |menu, project|
```
is this the first time a module adds a menu entry? Or why did you needed a new method for that?

ba1ash replied:
Yes, I needed a new method for that. It behaves differently from `menu` or `add_menu_item`. The difference is a possibility to depend on a project. for example.

Kharonus commented on modules/storages/spec/features/storages_menu_links_spec.rb:
```diff
@@ -0,0 +1,76 @@
+    expect(page).not_to have_link(storage1.name, href: storage1.host)
+  end
+
+  context 'when user is logged in without permissions to see storage links' do
```
rubocop is right here :D

Kharonus reviewed (commented):
I just have a few questions, looks very much done for me apart from those.

ba1ash committed:
Name context correctly.

ba1ash committed:
Put `external_link?` under `node_action_allowed?`.

Kharonus reviewed (approved):
LGTM

ba1ash committed:
Refactor code to be sematically better.

`external_link` is not appropriate name for this option.
Actually we would like to explicitly say that we are skipping a
permissions check and it does not matter if the link is external or internal.
